### ISSUE TITLE
Missing sub-module import statement is not detected

### ISSUE THREAD

**Issue by pylint-bot** [PyCQA/pylint#5]
Originally reported by: **Andrew Carman (BitBucket: [carmandrew](http://bitbucket.org/carmandrew), GitHub: @carmandrew?)**

---

When you run pylint on the following file, it reports no errors, but there is very clearly an error when running (AttributeError: 'module' object has no attribute 'saxutils').

```python
import xml.sax.handler

print(xml.sax.handler)
print(xml.sax.saxutils)

```

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/5

**Comment by gyermolenko** [comment]
I tried to run pylint 1.9.5 (python2.7) on files with
```
#!python

import xml.sax.handler
print xml.sax.handler
print xml.sax.saxutils
```
and
```
#!python

import os.path
print os.name
```
-  no AttributeError s. 

Maybe we can close the issue?

**Comment by PCManticore** [comment]
@gyermolenko You need to run `python` over those files and you'll see a `AttributeError: module 'xml.sax' has no attribute 'saxutils'` error. I think this still needs to be fixed.

**Comment by Pierre-Sassoulas** [comment]
This still exists in 2.9.0:

five.py
```python
import xml.sax.handler

print(xml.sax.handler)
print(xml.sax.saxutils)
```

pylint five.py:
```

------------------------------------
Your code has been rated at 10.00/10
```

python3 five.py:
```
<module 'xml.sax.handler' from '/usr/lib/python3.8/xml/sax/handler.py'>
Traceback (most recent call last):
  File "c.py", line 3, in <module>
    print(xml.sax.saxutils)
AttributeError: module 'xml.sax' has no attribute 'saxutils'
```

**Comment by kasium** [comment]
I also found this recently. After some debugging I found the source: https://github.com/PyCQA/pylint/blob/35254c29eb3a0f1c7847cfcb3451501a4180373d/pylint/checkers/variables.py#L1430

In short, if an import is found, only the first part (so e.g. `os`) is used for inferred, so that for all other checkers it seems, that `import os` was used, instead of `os.path`. Not sure if this can be changed easily

**Comment by Pierre-Sassoulas** [comment]
This seems promising @kasium.  Do you want to handle the oldest issue still opened ? 😄 I don't know the scope of this either. You can dig a little and just report your finding if this happens to be bigger than what you're prepared to do.

**Comment by kasium** [comment]
@Pierre-Sassoulas so my first investigation is not the root cause.

As far as I could track down the issue, if a submodule is loaded, the parent is loaded as well, so that a member check like `os.path` can succeed. However. since the whole `os` module is loaded, checks like `os.name` will succeed as well.

Instead of loading the whole `os` module if a child is requested, we maybe just want to return a wrapper, which only let's checks for `os.path` succeed but not `os.name`. Not sure, how complicated this will be

**Comment by Pierre-Sassoulas** [comment]
> if a submodule is loaded,

Do you mean by astroid or pylint ? Could you point to the specific code doing the loading ?

**Comment by clavedeluna** [comment]
Re-visiting old issues. Reproduced this with

```
# test.py
import xml.sax.handler

print(xml.sax.handler)
print(xml.sax.saxutils)
```

```
pylint test.py

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)
```

while Python output is

```
python test.py
<module 'xml.sax.handler' from '/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/xml/sax/handler.py'>
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    print(xml.sax.saxutils)
AttributeError: module 'xml.sax' has no attribute 'saxutils'
```
